### ISSUE TITLE
Reschedule or restart a task on health check failure

### ISSUE THREAD

**Issue by samart** [hashicorp/nomad#2973]
It would nice if a failing check restarted a task or re-scheduled it elsewhere.  

Consul provides a service discovery healthcheck, but when a service is unresponsive we'd like to restart it.   Our mesos clusters do this with marathon keep-alive healthchecks and it works well to keep applications responsive.

We should be able to specify at least:

- gracePeriod to wait before starting to healthcheck, after the task has started
- number of check failures before a restart/re-schedule 
- interval between checks
- timeout on the check attempt

Including the current nomad restart options would be nice as well.  Max restart attempts, mode, etc.

**Comment by epetrovich** [comment]
https://github.com/hashicorp/nomad/issues/876
This is an essential feature, but I've did it using my own script that is launched in the task group of the main app. Restarts of individual tasks/allocation  is possible via Consul-kv and template engine. I just wrote a simple script that checks the status of consul healthcheck and implements my own restart logic via Consul KV. 
Here's my working script for that. 
[watcher.py.gz](https://github.com/hashicorp/nomad/files/1204461/watcher.py.gz)
